test/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/SubsidyControllerSpec.scala

```diff
@@ -280,6 +282,272 @@ class SubsidyControllerSpec
+        }
+
+        "the user selects 'false' for reported payment returning user" in {
+          inSequence {
+            mockAuthWithEnrolmentAndValidEmail(eori1)
+            mockRetrieveUndertaking(eori1)(undertaking.some.toFuture)
+            mockGet[SubsidyJourney](eori1)(Right(journeyWithReportedPaymentReturningUser.some))
+            mockUpdate[SubsidyJourney](eori1)(
+              Right(journeyWithReportedPaymentReturningUser.setReportedPaymentReturningUser(false))
+            )
+          }
+
+          val result = performAction(
+            "report-payment-return" -> "false"
+          )
+
+          status(result) shouldBe SEE_OTHER
+          redirectLocation(result) should contain(routes.NoClaimNotificationController.getNoClaimNotification.url)
+        }
+      }
+    }
+
+    "handling request to get returning user page" must {
+
+      def performAction = controller.getReportedPaymentReturningUserPage(
+        FakeRequest(GET, routes.SubsidyController.getReportedPaymentReturningUserPage.url)
+      )
+
+      "call to get session fails" in {
+        inSequence {
+          mockAuthWithEnrolmentAndValidEmail()
+          mockRetrieveUndertaking(eori1)(undertaking.some.toFuture)
+          mockGetOrCreate[SubsidyJourney](eori1)(Left(ConnectorError(exception)))
+        }
+        assertThrows[Exception](await(performAction))
+      }
+
+      "display the page" when {
+        "a request is made" in {
+          inSequence {
+            mockAuthWithEnrolmentAndValidEmail()
+            mockRetrieveUndertaking(eori1)(undertaking.some.toFuture)
+            mockGetOrCreate[SubsidyJourney](eori1)(Right(subsidyJourney))
+          }
+          checkPageIsDisplayed(
+            performAction,
+            messageFromMessageKey("reportPaymentReturnJourney.title"),
+            { doc =>
+              doc.select(".govuk-back-link").attr("href") shouldBe routes.AccountController.getAccountPage.url
+              doc
+                .select("form")
+                .attr("action") shouldBe routes.SubsidyController.postReportedPaymentReturningUserPage.url
+            }
+          )
+        }
+      }
+    }
+
+    "handling get to reported no custom subsidy page" must {
+      def performAction(): Future[Result] = controller.getReportedNoCustomSubsidyPage(FakeRequest())
+
+      "render the page with updated form and correct URLs" in {
+        val journeyWithReportedNonCustomSubsidyReturningUser = subsidyJourney
+          .setReportedPaymentReturningUser(true)
+
+        inSequence {
+          mockAuthWithEnrolmentAndValidEmail(eori1)
+          mockRetrieveUndertaking(eori1)(undertaking.some.toFuture)
+          mockGetOrCreate[SubsidyJourney](eori1)(Right(journeyWithReportedNonCustomSubsidyReturningUser))
+        }
+
+        val result = performAction()
+
+        status(result) shouldBe OK
+        contentAsString(result) should include(
```

Comment:
i think its better to use JSoup to parse the html and then use an attribute (preferably the id) to select the element on the page and then you can check for things. You can add an id so that you can easily select it. We want to avoid just searching the returned string from contentAsString. Our tests should be a lot more deliberate.

Reply:
Yeah the jsoup stuff should also give better mismatch errors. 

For example got "meow" expected "mooo" where as the string stuff will say I cannot find the thingy in a big chunk of html making it a game of hunt the needle in the haystack. Id not found etc.